### ISSUE TITLE
Memory usage

### ISSUE THREAD

**Issue by wenbostar** [lazear/sage#33]
Is there a parameter to reduce the memory usage when the database is very large?

**Comment by lazear** [comment]
No, and this is a non-goal. I suggest using an instance with more RAM, or running multiple searches with smaller databases if needed.

Curious though - what's the database size and available memory?

**Comment by wenbostar** [comment]
This is for a phospho data using human swissprot database running on a colab node with 12G RAM. It works well on a non-phopho dataset.

```
{'database': {'bucket_size': 16384,
  'fragment_min_mz': 150.0,
  'fragment_max_mz': 1500.0,
  'enzyme': {'missed_cleavages': 1, 'cleave_at': 'KR', 'restrict': 'P'},
  'static_mods': {'C': 57.0216},
  'decoy_tag': 'rev_',
  'generate_decoys': True,
  'variable_mods': {'M': 15.9949, 'S': 79.99, 'T': 79.99, 'Y': 79.99}},
 'deisotope': True,
 'chimera': False,
 'max_fragment_charge': 1,
 'report_psms': 1,
 'precursor_tol': {'ppm': [-50, 50]},
 'fragment_tol': {'ppm': [-10, 10]},
 'isotope_errors': [-1, 3]
```

Below is the error message I got:
```
tcmalloc: large alloc 1360453632 bytes == 0x559fc4ac4000 @  0x7fd71cf8a1e7 0x559d28f30c95 0x559d28e96e8e 0x559d29280f38 0x559d28f0c78e 0x559d28ec3943 0x559d28f0db67 0x7fd71bfa6c87 0x559d28ea7f99
tcmalloc: large alloc 1360453632 bytes == 0x55a015c32000 @  0x7fd71cf8a1e7 0x559d2927f224 0x559d28f0c78e 0x559d28ec3943 0x559d28f0db67 0x7fd71bfa6c87 0x559d28ea7f99
```

**Comment by lazear** [comment]
Yeah that one will definitely required more memory 😄. I know MSFragger supports database splitting, so you could try it out if using a different node size isn't feasible.

Given that it's possible to rent EC2 instances with 128GB of ram for ~$1/hr, I decided that the complexity of implementing database splitting wasn't worth doing.

**Comment by wenbostar** [comment]
Got it. Thanks.

**Comment by jspaezp** [comment]
Hey there!

Do you think there is a way to "eyeball" the amount of memory required?
@wenbostar would you mind sharing how much memory was used on your phospho search?
Could mem usage on modification-containing searches be added to any of the benchmarks?

Best,
Sebastian!

**Comment by lazear** [comment]
@jspaezp in theory you could calculate the memory required for the fragment database - each fragment internally requires 8 bytes of ram, each peptide sequence has variable space requirements.

Digesting the latest human uniprot with the following mods (8 - 30 AA's): 
```json
      "static_mods": {
        "C": 57.0214
      },
      "variable_mods": {
        "M": 15.9949,
        "S": 79.966,
        "T": 79.966,
        "Y": 79.966
      },
```

Generates 649,470,153 fragments. Storing the fragments will cost ~4.8 GB + memory for storing peptides. I can complete generation & search on an M1 Air with 16 GB of ram. Peak memory usage clocked in around ~9.4 GB.

Currently only one benchmark measures memory consumption with variable mods:

<strong>-100 to +500 Da precursor, ±10 ppm fragment window with Met(Ox):</strong>
Engine | Runtime | Max RAM usage | PSMs at 1% FDR
-- | -- | -- | --
Sage | 68 s | 7.3 GB | 27,500
MSFragger | 340 s | 11.55 GB | 27,589
Improvement | 80% less runtime | 37% less RAM | 0.3% less PSMs

**Comment by fcyu** [comment]
@lazear The comparison with MSFragger is very interesting. May I ask some questions?

1. How many variable modification does Sage support on a single peptide?
2. Give a variable modification, what is the max occurrence on a single peptide?
3. Is there any max variable modification permutations limit in Sage? 
4. Have you set `localize_delta_mass = 1` for MSFragger? I don't think Sage supports localize-aware open search. For fair comparison, you should set it to 0 for MSFragger.
5. What is the `-Xmx` flag you gave to Java?

Thanks,

Fengchao

**Comment by lazear** [comment]
Hi @fcyu, you can see the parameters used here: https://github.com/lazear/sage/blob/master/figures/benchmark_params/fragger.open (For the Met-ox search I just un-commented the default variable modification search)

1. Unlimited
2. Currently 1
3. All permutations will be generated
4. `localize_delta_mass = 1` is set, since that was the recommended setting by FragPipe and per our emails. I will admit that I'm not a MSFragger expert (I've never used it before starting this) 😄. I can re-run this benchmark later with it set to 0. I have a branch for localization, but it's not finalized yet
5. I don't believe I set any memory limit - MSFragger was run on CLI.

To clarify points 1-3, we can use an example from the unit tests in [`sage/peptide.rs`](https://github.com/lazear/sage/blob/9445b5827a0c2d3b1418abf3b6880d7ceecfa19d/crates/sage/src/peptide.rs#L203). Modifying this to produce, for instance, peptides with 2 Met(Ox) instances is on my to-do list. In the case of the comparison of, the "Unlimited" options have no effect, since 

Apply Met(Ox) and Cys(CAM) variable mods to the peptide "GCMGCMG" will yield the following internal peptides.

```rs
"GC[+57]M[+16]GCMG",
"GCM[+16]GC[+57]MG",
"GCM[+16]GCMG",
"GC[+57]MGCM[+16]G",
"GCMGC[+57]M[+16]G",
"GCMGCM[+16]G",
"GC[+57]MGCMG",
"GCMGC[+57]MG",
"GCMGCMG",
```

If you have suggestions on parameters to tweak for MSFragger to align the benchmarks, please share them!

**Comment by lazear** [comment]
I will re-run - I'd like any comparisons to be as fair & good-faith as possible! These searches were run on an instance with 32 GB of ram, so presumably MSFragger using 12 GB of ram indicates it has access to more than >1/4 of memory (unless that's runtime overhead?)

For the missed_cleavages=2, I should've posted the whole snippet describing this benchmark on the github pages site. It looks like I just copied over the Sage params file
> I re-ran the same analysis as above, this time including methionine oxidation as a variable modification, and allowing for 2 missed cleavages to mimic a typical "real life" database search.

So both engines had cleavages set to 2. For this comparison, run time did not include Mokapot, which was used for FDR refinement for both. For the other comparisons (without variable mods), Mokapot time was included in the searches - but Mokapot is really fast in general for small searches like these, ~9 seconds IIRC. I chose Mokapot because it was easy to use results from MSFragger, Sage, Comet

**Comment by fcyu** [comment]
Due to the nature of Java, the `-XmxNg` matters for the memory usage and speed. I think you should set as large as possible as Sage is supposed to be able to use the whole free memory, like C++, right? 

On the other hand, if setting `-XmxNg` to a large value, the memory comparison will look bad for MSFragger since it will use as much memory as possible. To me, this is the amazing thing for MSFragger. When you have small memory, it slices the database and sacrifices the speed to get the job done. When you have big memory, it uses as much memory as possible to speed-up the searches.

Another suggestion (sorry that I have too many suggestions): it would be excellent if you can post the log from the comparison.

Best,

Fengchao

**Comment by lazear** [comment]
@fcyu check this out: https://gist.github.com/lazear/21c11a74a1849c987e8c465bd3ada784
I pasted the parameters followed by log.
`/usr/bin/time -v java -Xmx32g -jar MSFragger-3.5.jar fragger.open b1906_293T_proteinID_01A_QE3_122212.mzML > fragger.log2 2>&1`

Turning off localization makes a big difference in search speed!

Engine | Runtime | Max RAM usage
-- | -- | --
Sage | 65 s | 6.75 GB
MSFragger | 156 s | 9.7 GB

I agree that the database slicing is a nice feature, that's why I recommended MSFragger above! I also agree that memory usage isn't a big deal & shouldn't really be counted against a program in a benchmark. RAM is meant to be used!!

**Comment by wenbostar** [comment]
> @wenbostar are you running multiple files in parallel for this search? That might be the bigger issue on a memory constrained node. Sage will read num_cpus/2 files into memory, which could swamp memory constrained instances. You could try running it with the `--no-parallel` CLI flag or specify `"parallel": false` in the JSON configuration

No. Only one MS/MS file.
